README.md
<img src="https://s3.amazonaws.com/warriorbeatapp-hosting-mobilehub-1527922673/imgs/logo.jpg" align="right" height="170" width=170>

# WarriorBeatApi

Flask Zappa API for Warrior Beat</br>
Hosted via AWS

## Install & Setup

### Dependencies / Prerequisites

- [Python 3.6](https://www.python.org/)
- [Zappa](https://github.com/Miserlou/Zappa)
- Python Virtual Environment (_[Pipenv recommended](https://pipenv.readthedocs.io/en/latest/)_)
- [AWS CLI](https://aws.amazon.com/cli/)

### Setup

Clone the Repo

```sh
$ git clone https://github.com/WarriorBeat/WarriorBeatApi.git

$ cd WarriorBeatApi
```

Create and enter python virtual environment. Install dependencies.

```sh
$ pipenv shell
$ pipenv install --dev
$ pipenv sync
```

Configure AWS if you have not already.

```sh
$ aws configure
# Enter your IAM name and secret key
```

> [AWS Configure Docs](https://docs.aws.amazon.com/cli/latest/reference/configure/index.html)

The server is now installed and configured.

## Setting up your Environment

To setup a development/testing environment, you require the following:

1. Flask Instance Config
2. Flask Env Variables
3. A Local AWS DynamoDB Server
4. A Local AWS S3 Storage Server

Create an instance config file for flask.

```sh
# In WarriorBeatAPI folder
$ mkdir instance && cd instance
$ touch instance/config.py
```

In instance/config.py, configure as you wish. Here is a sample instance/config.py:

```python
# instance/config.py
DEBUG = True
SECRET_KEY = 'superdupersecretkey'
SERVER_NAME = 'localhost:5000'
```

In order to run the application, you need to set some environment variables:

```sh
$ export FLASK_APP=warriorbeat
$ export FLASK_ENV=development
# To make use of local database/storage:
$ export FLASK_TESTING=True
```

> NOTE: SERVER_NAME is required if you wish to use/work on webhooks in your testing environment.

The Local AWS servers can be configured in a few different ways, but in my opinion the easiest way is to use [Docker](https://www.docker.com/).

Once you have Docker installed, you need to pull the images and run the docker containers.

**DynamoDB:**

Pull the image:

```sh
$ docker pull amazon/dynamodb-local
```

> By default, this dynamodb image keeps all its data in memory. Therefore, by restarting or stopping the container, you effectively wipe the testing database.
> You can read more [here](https://hub.docker.com/r/amazon/dynamodb-local/).

Run the image:

```sh
$ docker run -p 8000:8000
```

Running the image like this doesn't detach it from your active terminal process, so to stop it simply `ctrl-c`

**S3 Storage:**

Pull the image:

```sh
# To store all files in memory:
$ docker pull scality/s3server:mem-latest
# If you want to keep the files:
$ docker pull scality/s3server:latest
```

> For my purposes, I generally just run the S3 Server in memory. You can read more [here](https://hub.docker.com/r/scality/s3server/).

Run the image:

```sh
$ docker run -d --name s3server -p 9000:8000 scality/s3server:mem-latest
```

To start/stop the image:

```sh
# start
$ docker start s3server
# stop
$ docker stop s3server
```

### Running the Server

Once you have setup your environment, running the server is pretty simple.

Make sure your environment variables are set (as in step two from above), you docker containers are up and running, and run:

```sh
$ flask run
```

Or, make a quick `run` bash script if your like me and forget to reset your environment variables when starting to work again:

```sh
#!/usr/bin/env bash
export FLASK_APP=warriorbeat
export FLASK_ENV=development
export FLASK_TESTING=True

flask run
```

Save the above as `run` in the projects root directory and set it as executable:

```sh
$ chmod +x ./run
```

Now you can run the server by calling the script:

```
$ ./run
```

## Running the Tests

Finally, to ensure that everything is setup properly, you need to run the UnitTests in the `tests/` directory.

> Make sure to have your Docker containers and flask server running.

Generally your IDE will come with functionality to recognize and run these tests, but you can also run them via `python -m unittest` command.

For VSCode, the Python Extension comes with full testing suite functionality. To get it started, open the action menu via: `#(your OS super key, CMD for macOS) ⌘CMD + P` and select the `> Python: Discover Unit Tests` command. Following that, run the `> Python: Run All Unit Tests` command.

You should now have a full development/testing environment ready to go.


CHANGELOG.md
# [](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.7.0...v) (2019-01-09)



# [0.7.0](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.6.5...v0.7.0) (2019-01-09)


### ✨ Features

* **Author:** Added Author Stats to Author Resource  ([#6](https://github.com/WarriorBeat/WarriorBeatApi/issues/6)) ([57e3d62](https://github.com/WarriorBeat/WarriorBeatApi/commit/57e3d62))
* **User:** Added User Attributes to Schema/Model ([07d2b14](https://github.com/WarriorBeat/WarriorBeatApi/commit/07d2b14))
* **User:** Added User Attributes to UserResource ([fcf06f6](https://github.com/WarriorBeat/WarriorBeatApi/commit/fcf06f6))


### 🐛 Bug Fixes

* Updated Dependencies ([328fefe](https://github.com/WarriorBeat/WarriorBeatApi/commit/328fefe))


### ♻️ Chores

* **release:** v0.7.0 ([2a4a201](https://github.com/WarriorBeat/WarriorBeatApi/commit/2a4a201))



## [0.6.5](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.6.4...v0.6.5) (2019-01-03)


### ✨ Features

* **User:** User Resource  ([#5](https://github.com/WarriorBeat/WarriorBeatApi/issues/5)) ([3fabcef](https://github.com/WarriorBeat/WarriorBeatApi/commit/3fabcef))


### 🐛 Bug Fixes

* Sentry loading in dev/debug environments ([88c3dfe](https://github.com/WarriorBeat/WarriorBeatApi/commit/88c3dfe))


### ♻️ Chores

* **release:** v0.6.5 ([0c1089f](https://github.com/WarriorBeat/WarriorBeatApi/commit/0c1089f))



## [0.6.4](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.6.3...v0.6.4) (2018-12-21)



## [0.6.3](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.6.2...v0.6.3) (2018-12-09)



## [0.6.2](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.6.1...v0.6.2) (2018-12-08)



## [0.6.1](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.6.0...v0.6.1) (2018-12-08)



# [0.6.0](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.5.0...v0.6.0) (2018-12-05)



# [0.5.0](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.4.0...v0.5.0) (2018-12-04)



## [2.1.1-beta](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.3.1...v2.1.1-beta) (2018-12-03)



# [2.1.0-beta](https://github.com/WarriorBeat/WarriorBeatApi/compare/v0.3.0...v2.1.0-beta) (2018-12-03)





warriorbeat/warriorbeat/utils/data.py
"""
    warriorbeat/api/utils/data.py
    Data Handlers
"""


import os

import boto3
import requests
from botocore.exceptions import ClientError

# Connection Info
TABLES = {
    'author': {
        'table_name': 'author-table',
        'primary_key': 'authorId'
    },
    'post': {
        'table_name': 'post-table',
        'primary_key': 'postId'
    },
    'media': {
        'table_name': 'media-table',
        'primary_key': 'mediaId'
    },
    'feedback': {
        'table_name': 'user-feedback-table',
        'primary_key': 'feedbackId'
    },
    'category': {
        'table_name': 'category-table',
        'primary_key': 'categoryId'
    },
    'poll': {
        'table_name': 'poll-table',
        'primary_key': 'pollId'
    },
    'user': {
        'table_name': 'user-table',
        'primary_key': 'userId'
    }
}

BUCKETS = {
    'media': {
        'bucket_name': 'media-bucket',
        'parent_key': 'media/'
    }
}

# Environment Variables
TESTING = os.environ['FLASK_TESTING']
AWS_DEV = os.environ['AWS_DEV']


class DynamoDB:
    """
    AWS Boto3 DynamoDB
    Handles transactions with database

    params:
    table_name: string
        name of table to access
    """

    def __init__(self, table):
        if TESTING == 'True':
            self.dynamodb = boto3.resource(
                'dynamodb', region_name='localhost', endpoint_url='http://localhost:8000')
        else:
            self.dynamodb = boto3.resource('dynamodb')
        self.table = TABLES[table]
        _tname = self.table['table_name']
        self.table['table_name'] = _tname if AWS_DEV != 'True' else _tname + '-dev'
        self.db = self.dynamodb.Table(self.table['table_name'])

    def clean_item(self, item):
        """prepare item for database insertion"""
        clean = item.copy()
        for k, v in item.items():
            if type(v) is str and len(v) <= 0:
                clean.pop(k)
        return clean

    def add_item(self, item):
        """adds item to database"""
        item = self.clean_item(item)
        self.db.put_item(Item=item)
        return item

    def delete_item(self, itemId):
        """deletes item from database"""
        self.db.delete_item(Key={
            self.table['primary_key']: itemId
        })
        return itemId

    def get_item(self, itemId):
        """retrieves an item from database"""
        try:
            resp = self.db.get_item(Key={
                self.table['primary_key']: itemId
            })
            return resp.get('Item')
        except ClientError as e:
            print(e)
            return None

    def exists(self, itemId):
        """checks if an item exists in the database"""
        item = self.get_item(itemId)
        return False if item is None else item

    @property
    def all(self):
        """list all items in the database table"""
        resp = self.db.scan()
        items = resp["Items"]
        return items


class S3Storage:
    """
    AWS S3 Bucket
    Handles transactions with s3 Buckets

    params:
    bucket_name: string
        name of bucket to access
    """

    FILE_EXTENSIONS = {
        "image/jpeg": ".jpeg",
        "image/png": ".png"
    }

    def __init__(self, bucket):
        if TESTING == 'True':
            self.s3bucket = boto3.resource(
                's3', region_name='localhost', endpoint_url='http://localhost:9000', aws_access_key_id='accessKey1', aws_secret_access_key='verySecretKey1')
            self.s3client = boto3.client(
                's3', region_name='localhost', endpoint_url='http://localhost:9000', aws_access_key_id='accessKey1', aws_secret_access_key='verySecretKey1')
        else:
            self.s3bucket = boto3.resource('s3')
            self.s3client = boto3.client('s3')
        self.bucket = BUCKETS[bucket]
        self.bucket['bucket_name'] = self.bucket['bucket_name'] if AWS_DEV != 'True' else self.bucket['bucket_name'] + '-dev'
        self.storage = self.s3bucket.Bucket(self.bucket['bucket_name'])
        self.key = self.bucket['parent_key']

    def get_url(self, key, **kwargs):
        """generates url where the image is hosted"""
        aws_root = "https://s3.amazonaws.com"
        if TESTING == 'True':
            aws_root = "http://localhost:9000"
        url = f"{aws_root}/{self.bucket['bucket_name']}/{key}"
        return url

    def upload(self, path, key=None):
        """uploads a file to the s3 bucket"""
        if key:
            key = self.key + key
            self.storage.upload_file(path, key, ACL='public-read')
            return self.get_url(key)
        else:
            self.storage.upload_file(path, self.key, ACL='public-read')
            return self.get_url(self.key)

    def upload_obj(self, obj: tuple, key=''):
        """upload file object"""
        _key = self.key + key + obj[1]
        self.storage.put_object(Key=_key, Body=obj[0], ACL='public-read')
        return self.get_url(_key)

    def upload_from_url(self, url, **kwargs):
        """upload file from url"""
        img_stream = requests.get(url, stream=True)
        content_type = img_stream.headers.get('content-type', 'image/jpeg')
        file_ext = self.FILE_EXTENSIONS[content_type]
        img_obj = img_stream.raw
        img_data = img_obj.read()
        return (self.upload_obj((img_data, file_ext), **kwargs), file_ext)
